Client backends for the remote execution of circuits

BrunoLiegiBastonLiegi opened the pull request:
This PR add support for the  `QiboClientBackend` and a `QiskitClientBackend` implemented in [https://github.com/qiboteam/qibo-cloud-backends](https://github.com/qiboteam/qibo-cloud-backends) that allow for the remote execution of circuits. Furthermore, support for QASM `gate` command is added to the parser. 

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

scarrazza commented:
@BrunoLiegiBastonLiegi thanks for this, however let me suggest to move these backends directly in https://github.com/qiboteam/qibo-cloud-backends, so we avoid mixing the core code with external providers.

BrunoLiegiBastonLiegi commented:
Ah ok, sorry about that, I'll move them.

alecandido commented on src/qibo/backends/__init__.py:
```diff
@@ -8,7 +8,7 @@
 
 
-def construct_backend(backend, platform=None, runcard=None):
+def construct_backend(backend, platform=None, runcard=None, token=None, provider=None):
```
Maybe we should consider using 
```suggestion
def construct_backend(backend, **kwargs):
```

What do you think @scarrazza?

alecandido commented on src/qibo/backends/__init__.py:
```diff
@@ -41,7 +41,18 @@ def construct_backend(backend, platform=None, runcard=None):
             else:
                 platform = construct_backend(platform)
         return CliffordBackend(platform)
+    elif backend == "qibo":  # pragma: no cover
```
Shouldn't we call it *cloud*, or something like?

`"qibo"` is not much informative (in this context...)

BrunoLiegiBastonLiegi replied:
I originally called them `qibo-client` and `qiskit-client` but then reverted them to just `qibo` and `qiskit`. Maybe `qibo-cloud` and `qiskit-cloud`?

alecandido commented on src/qibo/backends/__init__.py:
```diff
@@ -83,13 +94,17 @@ def __new__(cls):
 
     @classmethod
-    def set_backend(cls, backend, platform=None, runcard=None):  # pragma: no cover
+    def set_backend(
+        cls, backend, platform=None, runcard=None, token=None, provider=None
+    ):  # pragma: no cover
```
Same proposal of above
```suggestion
    def set_backend(cls, backend, **kwargs):  # pragma: no cover
```

alecandido commented on src/qibo/backends/__init__.py:
```diff
@@ -127,8 +142,8 @@ def get_backend():
 
-def set_backend(backend, platform=None, runcard=None):
-    GlobalBackend.set_backend(backend, platform, runcard)
+def set_backend(backend, platform=None, runcard=None, token=None, provider=None):
```
`**kwargs` once more.

alecandido commented on tests/test_quantum_info_clifford.py:
```diff
@@ -107,6 +107,10 @@ def test_clifford_to_circuit(backend, nqubits, algorithm):
             symplectic_matrix_compiled, engine=backend
         ).symplectic_matrix
 
+        print(symplectic_matrix_from_symplectic == symplectic_matrix_original)
+        print(
+            (~symplectic_matrix_from_symplectic == symplectic_matrix_original).nonzero()
+        )
```
Debug prints?

You should drop them before merging.

alecandido commented on src/qibo/models/circuit.py:
```diff
@@ -1249,13 +1249,13 @@ def read_args(args):
-            line for line in qasm_code.split("\n") if line and line[:2] != "//"
-        )
-        lines = (line for line in lines.split(";") if line)
+        lines = [line for line in qasm_code.split("\n") if line and line[:2] != "//"]
```
```suggestion
        lines = [line for line in qasm_code.splitlines() if line and line[:2] != "//"]
```
More portable :)

alecandido commented on src/qibo/models/circuit.py:
```diff
@@ -1265,11 +1265,14 @@ def read_args(args):
+        composite_gates = {}  # composite gates created with the "gate" command
+
+        def parse_line(line):
+            line = line.replace(r"/\s\s+/g", " ").replace(r"^[\s]+", "")
```
```suggestion
            line = line.replace(r"/\s{2,}/g", " ").replace(r"^[\s]+", "")
```

But I really believe we should avoid building a parser procedurally with regexes...

alecandido reviewed (commented):
Despite the parser being the major upgrade, I didn't pay much attention to it, because we should use a better strategy for defining it (either using the official parser or make our own one, with a specification of a subset of the formal language used).

Using regexes is a simple option for a quick parser, but when things grow and last long it becomes quite involved and complex (and thus much harder to review...).

BrunoLiegiBastonLiegi commented:
Thanks @alecandido, I agree the parser has to be replaced with something more controllable. If the execution through cloud functionality is not urgent, we could delay this and try to incorporate the use of [https://github.com/openqasm/openqasm](https://github.com/openqasm/openqasm) as we discussed separately.

scarrazza commented:
@BrunoLiegiBastonLiegi do you prefer to merge this now or wait until qibo-cloud-backens is released? Openqasm could be further upgraded in a separate PR.

BrunoLiegiBastonLiegi commented:
I don't know, the only additional thing this PR is providing is the support to the qasm gate command. So it depends, if that is something that is nice to have without waiting for the cloud backends we could merge this.  Otherwise we could wait.

scarrazza commented:
Ok, lets wait until we test and release the cloud backends.

scarrazza commented:
Meanwhile, could you please fix tests?

BrunoLiegiBastonLiegi commented:
Ok I added a reference in the `qibo` api reference docs about `qibo-cloud-backends`, linking to the official documentation (which also contains usage examples). I also added an optional `cloud` group in poetry.
Both this and `qibo-cloud-backends` PR should be ready, we just need to release `qibo-cloud-backends-0.0.1` and then merge this I guess.

alecandido commented on pyproject.toml:
```diff
@@ -93,6 +93,12 @@ cupy-cuda12x = "^12.0.0"
 cuquantum-python-cu12 = "^23.3.0"
 qibojit = { git = "https://github.com/qiboteam/qibojit.git" }
 
+[tool.poetry.group.cloud]
+optional = true
+
+[tool.poetry.group.cloud.dependencies]
+qibo-cloud-backends = { git = "https://github.com/qiboteam/qibo-cloud-backends.git" }
+
```
Instead of group (that is a Poetry feature, not passing to the package) you should use a pip extra, let's say cloud, such that it will be possible to install with:
```sh
pip install qibo[cloud]
```
https://python-poetry.org/docs/pyproject/#extras

In practice, just move the dependency up, and mark it with `optional = true`. Then, add the extra group, (here it has been never done, and even the other groups are mostly abusing this Poetry feature, i.e. all the cuda groups...).
If you want to have an example you could also have a look to Qibolab and Qibocal.

alecandido replied:
Solved in 1d12b45

alecandido commented:
@BrunoLiegiBastonLiegi any known blocker for this?

I believe @MatteoRobbiati could be busy, so I'm adding @stavros11 for the review (and @scarrazza in cc).

scarrazza commented on doc/source/api-reference/qibo.rst:
```diff
@@ -2414,3 +2414,9 @@ Alternatively, a Clifford circuit can also be executed starting from the :class:
+Cloud Backends
+^^^^^^^^^^^^^^
+
+Additional backends, that support the remote execution of quantum circuits through cloud service providers, are provided by the optional qibo plugin `qibo-cloud-backends <https://github.com/qiboteam/qibo-cloud-backends>`_. For more information please refer to the `official documentation <https://qibo.science/qibo-cloud-backends/stable/>`_.
```
```suggestion
Additional backends, that support the remote execution of quantum circuits through cloud service providers, such as IBM and QRC-TII, are provided by the optional qibo plugin `qibo-cloud-backends <https://github.com/qiboteam/qibo-cloud-backends>`_. For more information please refer to the `official documentation <https://qibo.science/qibo-cloud-backends/stable/>`_.
```

alecandido commented on src/qibo/backends/__init__.py:
```diff
@@ -36,11 +36,7 @@ def construct_backend(backend, **kwargs):
-        if platform in ("cupy", "numba", "cuquantum"):
-            platform = construct_backend("qibojit", platform=platform)
-        elif platform == "numpy":
-            platform = construct_backend(platform)
```
@BrunoLiegiBastonLiegi I'm not sure why you removed this...

alecandido replied:
I assumed it was an error due to merging an old version of the `master` branch, and I reintroduced it.

If I was wrong, just let me know.

BrunoLiegiBastonLiegi replied:
Mmmh yes I think there was some mess due to the master merge, but I don't think is needed anymore as now we only pass the name of the engine as a string. At least in master it's like that.

alecandido replied:
Ok, then I converted `platform` in `kwargs["platform"]`. Maybe it was like that, and somehow changed during the merge.

(I only had a look to the diff, so I did not realize before, sorry)

alecandido commented:
✅ great to see you again!

alecandido commented:
I rebased on `master`, such that it can be cleanly merged (#1236 was merged in the meanwhile).

Whenever tests are passing, I'd merge this @BrunoLiegiBastonLiegi
